File: assets/reports-model/Donor Blood Type Distribution.py
"""
Donor Blood Type Distribution Interactive Chart
================================================
Creates an interactive bar chart showing donor blood type distribution.

Definition (to align with active donor universe):
- Use the eligibility table
- Work from the LATEST eligibility row per donor_id
- Include ONLY donors whose latest status is approved/eligible
- Count each donor_id ONCE (no duplicates)
"""

import json
import sys
from typing import Dict, List

import plotly.graph_objects as go

from database import DatabaseConnection
from config import BLOOD_TYPES


# Color palette - red shades for blood types (matching R version)
COLORS = ["#dc2626", "#ef4444", "#f87171", "#fca5a5",
          "#991b1b", "#b91c1c", "#dc2626", "#ef4444"]


def _parse_dt(value):
    from datetime import datetime

    if not value:
        return None
    try:
        text = str(value)
        if "Z" in text:
            return datetime.fromisoformat(text.replace("Z", "+00:00"))
        return datetime.fromisoformat(text)
    except Exception:
        return None


def fetch_approved_latest_blood_types() -> List[Dict]:
    """
    Fetch one blood_type per donor based on the LATEST eligibility row
    whose status is approved/eligible and blood_type is not null.

    Returns a list of records like: { donor_id, blood_type } with
    each donor_id appearing at most once.
    """
    db = DatabaseConnection()
    records: List[Dict] = []

    if db.connect():
        try:
            endpoint = (
                "eligibility?"
                "select=eligibility_id,donor_id,status,blood_type,created_at"
                "&blood_type=not.is.null"
                "&order=donor_id.asc,created_at.asc"
            )
            records = db.supabase_request(endpoint)
        finally:
            db.disconnect()

    latest_by_donor: Dict[int, Dict] = {}

    for rec in records:
        donor_id = rec.get("donor_id")
        if donor_id is None:
            continue
        created = _parse_dt(rec.get("created_at"))
        existing = latest_by_donor.get(donor_id)
        if existing is None:
            latest_by_donor[donor_id] = {
                "status": rec.get("status"),
                "blood_type": rec.get("blood_type"),
                "_created": created,
            }
        else:
            prev_created = existing.get("_created")
            if created and (prev_created is None or created > prev_created):
                latest_by_donor[donor_id] = {
                    "status": rec.get("status"),
                    "blood_type": rec.get("blood_type"),
                    "_created": created,
                }

    result: List[Dict] = []
    for donor_id, info in latest_by_donor.items():
        status_text = (info.get("status") or "").lower()
        blood_type = info.get("blood_type")
        if not blood_type:
            continue
        if "approved" in status_text or status_text == "eligible":
            result.append({"donor_id": donor_id, "blood_type": blood_type})

    return result


def aggregate_by_blood_type(donor_records: List[Dict]) -> List[Dict]:
    """
    Aggregate donors by blood type.
    
    Returns:
        List of dicts with blood_type, count, and percentage
    """
    blood_type_counts: Dict[str, int] = {bt: 0 for bt in BLOOD_TYPES}

    for record in donor_records:
        blood_type = record.get("blood_type")

        if blood_type and blood_type in BLOOD_TYPES:
            blood_type_counts[blood_type] += 1
    
    total = sum(blood_type_counts.values())
    
    # Build result list in order: O+, O-, A+, A-, B+, B-, AB+, AB-
    blood_type_order = ["O+", "O-", "A+", "A-", "B+", "B-", "AB+", "AB-"]
    result = []
    
    for bt in blood_type_order:
        count = blood_type_counts.get(bt, 0)
        percentage = round((count / total) * 100, 1) if total > 0 else 0
        result.append({
            "blood_type": bt,
            "count": count,
            "percentage": percentage
        })
    
    return result


def calculate_rh_summary(donor_data: List[Dict]) -> List[Dict]:
    """Calculate Rh factor summary (Positive vs Negative)."""
    positive_count = sum(d["count"] for d in donor_data if d["blood_type"].endswith("+"))
    negative_count = sum(d["count"] for d in donor_data if d["blood_type"].endswith("-"))
    total = positive_count + negative_count
    
    return [
        {
            "rh_factor": "Positive",
            "count": positive_count,
            "percentage": round((positive_count / total) * 100, 1) if total > 0 else 0
        },
        {
            "rh_factor": "Negative",
            "count": negative_count,
            "percentage": round((negative_count / total) * 100, 1) if total > 0 else 0
        }
    ]


def create_bar_chart(donor_data: List[Dict], total_donors: int) -> go.Figure:
    """Create an interactive bar chart using Plotly."""
    blood_types = [d["blood_type"] for d in donor_data]
    counts = [d["count"] for d in donor_data]
    
    fig = go.Figure(data=[
        go.Bar(
            x=blood_types,
            y=counts,
            marker=dict(
                color=COLORS[:len(blood_types)],
                line=dict(color="rgb(139, 0, 0)", width=1.5)
            ),
            hovertemplate=(
                "Blood Type: %{x}<br>"
                "Count: %{y}<br>"
                "<extra></extra>"
            )
        )
    ])
    
    fig.update_layout(
        title=dict(
            text=f"<b>Donor Blood Type Distribution</b><br><sub>Total Donors: {total_donors}</sub>",
            font=dict(size=20)
        ),
        xaxis=dict(
            title="<b>Blood Type</b>",
            tickfont=dict(size=12)
        ),
        yaxis=dict(
            title="<b>Number of Donors</b>",
            tickfont=dict(size=12),
            gridcolor="#e5e7eb"
        ),
        plot_bgcolor="#f8fafc",
        paper_bgcolor="#ffffff",
        hovermode="closest",
        hoverlabel=dict(
            bgcolor="white",
            font_size=12,
            font_color="black",
            bordercolor="#e5e7eb"
        ),
        margin=dict(t=80, b=60, l=60, r=30)
    )
    
    return fig


def get_donor_blood_type_distribution_data() -> Dict:
    """
    Main function to get donor blood type distribution data.
    Returns a dictionary with the data for API consumption.
    """
    donor_records = fetch_approved_latest_blood_types()

    if not donor_records:
        return {
            "success": False,
            "message": "No active donors with blood type data found in eligibility table",
            "data": [],
            "total_donors": 0,
            "rh_summary": [],
        }

    donor_data = aggregate_by_blood_type(donor_records)
    total_donors = sum(d["count"] for d in donor_data)
    rh_summary = calculate_rh_summary(donor_data)
    
    return {
        "success": True,
        "data": donor_data,
        "total_donors": total_donors,
        "rh_summary": rh_summary
    }


def generate_chart_html(output_path: str = None) -> str:
    """Generate and optionally save the chart as HTML.

    IMPORTANT: Always create an HTML asset file, even when there is no data,
    so that forecast-asset.php never returns a 404. In the no-data case,
    we write a simple placeholder HTML with a friendly message.
    """
    try:
        donor_records = fetch_approved_latest_blood_types()

        if not donor_records:
            placeholder_html = """<!DOCTYPE html>
<html><head><meta charset="utf-8"><title>Donor Blood Type Distribution</title></head>
<body style="font-family: Arial, sans-serif; padding: 16px;">
  <h3 style="margin-top: 0;">Donor Blood Type Distribution</h3>
  <p>No donor blood type data is currently available. Please check back later.</p>
</body></html>"""
            if output_path:
                with open(output_path, "w", encoding="utf-8") as f:
                    f.write(placeholder_html)
                print(f"Placeholder blood type chart saved to: {output_path}", file=sys.stderr)
            return placeholder_html

        donor_data = aggregate_by_blood_type(donor_records)
        total_donors = sum(d["count"] for d in donor_data)

        fig = create_bar_chart(donor_data, total_donors)

        if output_path:
            fig.write_html(output_path)
            print(f"Chart saved to: {output_path}", file=sys.stderr)

        return fig.to_html(include_plotlyjs="cdn")
    except Exception as exc:
        placeholder_html = f"""<!DOCTYPE html>
<html><head><meta charset="utf-8"><title>Donor Blood Type Distribution</title></head>
<body style="font-family: Arial, sans-serif; padding: 16px;">
  <h3 style="margin-top: 0;">Donor Blood Type Distribution</h3>
  <p>An error occurred while generating this chart. Please check server logs.</p>
  <pre style="font-size: 11px; color: #6b7280;">{str(exc)}</pre>
</body></html>"""
        if output_path:
            try:
                with open(output_path, "w", encoding="utf-8") as f:
                    f.write(placeholder_html)
                print(f"Error placeholder blood type chart saved to: {output_path}", file=sys.stderr)
            except Exception:
                pass
        return placeholder_html


def print_summary():
    """Print a formatted summary to console (matching R output style)."""
    result = get_donor_blood_type_distribution_data()
    
    print("\n=== DONOR BLOOD TYPE DISTRIBUTION SUMMARY ===\n")
    print(f"Total Donors: {result['total_donors']}\n")
    
    for item in result["data"]:
        print(f"Blood Type {item['blood_type']}: {item['count']} donors ({item['percentage']}%)")
    
    # Rh factor summary
    print("\n=== RH FACTOR SUMMARY ===\n")
    for item in result["rh_summary"]:
        print(f"{item['rh_factor']}: {item['count']} donors ({item['percentage']}%)")


if __name__ == "__main__":
    import argparse
    
    parser = argparse.ArgumentParser(description="Donor Blood Type Distribution Report")
    parser.add_argument("--json", action="store_true", help="Output as JSON")
    parser.add_argument("--html", type=str, help="Save chart as HTML to specified path")
    args = parser.parse_args()
    
    if args.html:
        generate_chart_html(args.html)
    elif args.json:
        result = get_donor_blood_type_distribution_data()
        print(json.dumps(result, indent=2))
    else:
        print_summary()

File: assets/reports-model/R/Donor Blood Type Distribution.R
# ============================================
# Donor Blood Type Distribution Interactive Chart
# ============================================

set.seed(123)  # For reproducible sample data
# Common blood type distribution in Philippines
sample_blood_types <- c(
  rep("O+", 285),
  rep("A+", 210),
  rep("B+", 158),
  rep("AB+", 52),
  rep("O-", 35),
  rep("A-", 28),
  rep("B-", 21),
  rep("AB-", 15)
)

# Shuffle for randomness
sample_blood_types <- sample(sample_blood_types)

# Create a data frame (this simulates your donor_form data)
donor_data_raw <- data.frame(
  donor_id = 1:length(sample_blood_types),
  blood_type = sample_blood_types
)

# ============================================
# TO USE YOUR ACTUAL DATA:
# Read your data from CSV or other source
# ============================================
# Option 1: If you have a CSV file
# donor_data_raw <- read.csv("donor_form.csv")

# Option 2: If you have an Excel file
# install.packages("readxl")
# library(readxl)
# donor_data_raw <- read_excel("donor_form.xlsx")

# Option 3: If you manually input data
# donor_data_raw <- data.frame(
#   donor_id = c(1, 2, 3, ...),
#   blood_type = c("O+", "A+", "B+", ...)
# )

# ============================================
# Data Processing - Count by Blood Type
# ============================================
donor_data <- donor_data_raw %>%
  group_by(blood_type) %>%
  summarise(count = n()) %>%
  mutate(
    percentage = round((count / sum(count)) * 100, 1),
    blood_type = factor(blood_type, levels = c("O+", "O-", "A+", "A-", "B+", "B-", "AB+", "AB-"))
  ) %>%
  arrange(blood_type)

# Total donors
total_donors <- sum(donor_data$count)

# ============================================
# Create Interactive Bar Chart
# ============================================
# Define color palette - alternating red shades for positive/negative
colors <- c('#dc2626', '#ef4444', '#f87171', '#fca5a5', 
            '#991b1b', '#b91c1c', '#dc2626', '#ef4444')

# Create the plot
fig <- plot_ly(
  data = donor_data,
  x = ~blood_type,
  y = ~count,
  type = 'bar',
  marker = list(
    color = colors[1:nrow(donor_data)],
    line = list(color = 'rgb(139, 0, 0)', width = 1.5)
  ),
  hovertext = ~paste0(
    'Blood Type: ', blood_type, '<br>',
    'Count: ', count, '<br>',
    'Percentage: ', percentage, '%'
  ),
  hoverinfo = 'text',
  textposition = 'none'
) %>%
  layout(
    title = list(
      text = paste0(
        '<b>Donor Blood Type Distribution</b><br>',
        '<sub>Total Donors: ', total_donors, '</sub>'
      ),
      font = list(size = 20)
    ),
    xaxis = list(
      title = '<b>Blood Type</b>',
      tickfont = list(size = 12),
      titlefont = list(size = 14)
    ),
    yaxis = list(
      title = '<b>Number of Donors</b>',
      tickfont = list(size = 12),
      titlefont = list(size = 14),
      gridcolor = '#e5e7eb'
    ),
    plot_bgcolor = '#f8fafc',
    paper_bgcolor = '#ffffff',
    hovermode = 'closest',
    hoverlabel = list(
      bgcolor = 'white',
      font = list(size = 12, color = 'black'),
      bordercolor = '#e5e7eb'
    ),
    margin = list(t = 80, b = 60, l = 60, r = 30)
  ) %>%
  config(
    displayModeBar = TRUE,
    modeBarButtonsToRemove = c('lasso2d', 'select2d'),
    displaylogo = FALSE
  )

# Display the chart
fig

# ============================================
# Print Summary Statistics
# ============================================
cat("\n=== DONOR BLOOD TYPE DISTRIBUTION SUMMARY ===\n\n")
cat(sprintf("Total Donors: %d\n\n", total_donors))

for(i in 1:nrow(donor_data)) {
  cat(sprintf(
    "Blood Type %s: %d donors (%.1f%%)\n",
    donor_data$blood_type[i],
    donor_data$count[i],
    donor_data$percentage[i]
  ))
}

# ============================================
# Additional Analysis - Positive vs Negative
# ============================================
rh_summary <- donor_data_raw %>%
  mutate(
    rh_factor = ifelse(grepl("\\+", blood_type), "Positive", "Negative")
  ) %>%
  group_by(rh_factor) %>%
  summarise(count = n()) %>%
  mutate(percentage = round((count / sum(count)) * 100, 1))

cat("\n=== RH FACTOR SUMMARY ===\n\n")
for(i in 1:nrow(rh_summary)) {
  cat(sprintf(
    "%s: %d donors (%.1f%%)\n",
    rh_summary$rh_factor[i],
    rh_summary$count[i],
    rh_summary$percentage[i]
  ))
}
